src/olympia/addons/serializers.py

```diff
@@ -898,6 +904,30 @@ def validate(self, data):
                 )
         return data
 
+    def _save_icon(self, uploaded_icon):
+        # This is only used during update. For create it's impossible to send the icon
+        # as formdata while also sending json for `version`.
+        destination = os.path.join(self.instance.get_icon_dir(), str(self.instance.id))
+        if uploaded_icon:
+            original = f'{destination}-original.png'
+
+            storage = SafeStorage(user_media='addon_icons')
+            with storage.open(original, 'wb') as original_file:
+                for chunk in uploaded_icon.chunks():
+                    original_file.write(chunk)
+
+            self.instance.update(icon_type=uploaded_icon.content_type)
+            # note: we're calling the task without .delay() to resize immediately
```

Comment:
Why ?

Reply:
we return the links to the resized image files in the serialized response so I thought the files should actually exist at that point or they'd 404 if they were loaded immediately (or there was celery task queue traffic).  It should be quick enough to resize a single file to 32, 64, and 128.

Reply:
It might be quick enough in most cases, but it also calls `pngcrush_image`, which in turn calls a (potentially slow) binary, something I think we should avoid doing in a web process.

Current devhub works around the same problem in 2 ways:
- Before you click save, the preview is generated locally
- After you click save, it calls the icon with a `modified=` parameter in a loop until they stop 404ing.

This might not be super elegant but I think it's still better than executing the resize+crush code in a web process.

Reply:
I know, but how's that going to work with the api though?  Or are you suggesting we just return the urls as potentially 404s and expect the client to know what to do (and/or document how devhub does it in the api docs as a suggestion)?

Reply:
Yes, I think we can return the URLs as potentially 404s. They should be served without caching (and if we decide to add caching to errors, we could do a very short expiration) so they shouldn't even need devhub's trick and can use the URL as-is.